File: clone_graph.cpp
/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
    unordered_map<int, UndirectedGraphNode *> existed;
    UndirectedGraphNode * solveCloneGraph(UndirectedGraphNode * node)
    {
        UndirectedGraphNode * currNode = new UndirectedGraphNode(node -> label);
        existed[node -> label] = currNode;
        for (int i = 0; i < (node -> neighbors).size(); i++)
        {
            if (existed.find(node->neighbors[i] -> label) == existed.end())
            {
                currNode -> neighbors.push_back(solveCloneGraph((node -> neighbors)[i]));
            }
            else
            {
                currNode -> neighbors.push_back(existed[(node->neighbors[i]) -> label]);
            }
        }
        return currNode;
    }
public:
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        existed.clear();
        if (node == NULL) return NULL;
        return solveCloneGraph(node);
    }
};

File: Java/clone_graph.java
/**
 * Definition for undirected graph.
 * class UndirectedGraphNode {
 *     int label;
 *     ArrayList<UndirectedGraphNode> neighbors;
 *     UndirectedGraphNode(int x) { label = x; neighbors = new ArrayList<UndirectedGraphNode>(); }
 * };
 */
public class Solution {
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        if (node == null)
        {
            return null;
        }
        UndirectedGraphNode newHead = new UndirectedGraphNode(node.label);
        LinkedList<UndirectedGraphNode> queue = new LinkedList<UndirectedGraphNode>();
        HashMap<UndirectedGraphNode, UndirectedGraphNode> hashMap = new HashMap<UndirectedGraphNode, UndirectedGraphNode>();
        queue.push(node);
        hashMap.put(node, newHead);
        while(!queue.isEmpty())
        {
            UndirectedGraphNode tmpNode = queue.pop();
            for (int i = 0; i < tmpNode.neighbors.size(); i++)
            {
                if (hashMap.containsKey(tmpNode.neighbors.get(i)))
                {
                    hashMap.get(tmpNode).neighbors.add(hashMap.get(tmpNode.neighbors.get(i)));
                }
                else
                {
                    UndirectedGraphNode newNode = new UndirectedGraphNode(tmpNode.neighbors.get(i).label);
                    hashMap.put(tmpNode.neighbors.get(i), newNode);
                    hashMap.get(tmpNode).neighbors.add(newNode);
                    queue.push(tmpNode.neighbors.get(i));
                }
            }
        }
        return newHead;
    }
}
